### PR TITLE
[ALLUXIO-2226] Add a LocalFirstPolicy that without evict action

### PR DESCRIPTION
**Pull request by gjhkael**
https://alluxio.atlassian.net/browse/ALLUXIO-2226

### CODE REVIEW

**Comment by gjhkael** [comment]
alluxio-bot, check this please

File: core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * Tests that the local host is returned first.
+   */
+  @Test
+  public void getLocalFirst() {
+    String localhostName = NetworkAddressUtils.getLocalHostName();
+    LocalFirstWithoutEvictionPolicy policy = new LocalFirstWithoutEvictionPolicy();
+    List<BlockWorkerInfo> workerInfoList = new ArrayList<>();
+    workerInfoList.add(new BlockWorkerInfo(new WorkerNetAddress().setHost("worker1")
+        .setRpcPort(PORT).setDataPort(PORT).setWebPort(PORT), Constants.GB, 0));
```
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java]: We don't need to set these ports, they will be zero by default which is fine for this test

File: core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * Tests that local host is picked if none of the workers has enough availability.
+   */
+  @Test
+  public void getLocalWhenNoneHasSpace() {
+    String localhostName = NetworkAddressUtils.getLocalHostName();
+    LocalFirstWithoutEvictionPolicy policy = new LocalFirstWithoutEvictionPolicy();
+    List<BlockWorkerInfo> workerInfoList = new ArrayList<>();
+    workerInfoList.add(new BlockWorkerInfo(new WorkerNetAddress().setHost("worker1")
+        .setRpcPort(PORT).setDataPort(PORT).setWebPort(PORT), Constants.GB, 0));
```
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java]: It looks like worker1 has 1GB available space, so it's confusing that it's not allowed to allocate a 1GB block.

File: docs/_data/table/en/user-configuration.yml [hunk 3]
```diff
@@ -30,6 +30,9 @@ alluxio.user.file.worker.client.threads:
   How many threads to use for file worker clients to read from workers.
 alluxio.user.file.write.location.policy.class:
   The default location policy for choosing workers for writing a file's blocks
+alluxio.user.file.write.capacity.reserved.ratio:
```
Comment (aaudiber) [inline docs/_data/table/en/user-configuration.yml]: Why is this property useful? We don't use anything like this for the regular `LocalFirstPolicy` so it seems strange to use it here. Is it possible to break adding this property into a separate PR so that it doesn't block merging the new location policy?

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicy.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.List;
+
+import javax.annotation.concurrent.ThreadSafe;
+/**
+ * A policy that returns local host first, and if the local worker doesn't have enough availability,
+ * it randomly picks a worker from the active workers list for each block write.
+ * If No worker meets the demands, return local host.
+ */
+@ThreadSafe
+public class LocalFirstWithoutEvictionPolicy implements FileWriteLocationPolicy {
```
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicy.java]: This policy could still cause eviction when all workers are out of space, so it might be better to name it `LocalFirstAvoidEvictionPolicy`

File: core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java (hunk 2 above)
Comment (gjhkael) [inline core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java]: Because the policy is local first, even if worker1 has the available space but local will be choice first.

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicy.java (hunk 4 above)
Comment (gjhkael) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicy.java]: Seems quite reasonable.

File: docs/_data/table/en/user-configuration.yml (hunk 3 above)
Comment (gjhkael) [inline docs/_data/table/en/user-configuration.yml]: The property is discussed in this pull request(https://github.com/Alluxio/alluxio/pull/4423), Because my git commit history  is corrupted, So, I open the  new pull request.

File: core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java (hunk 2 above)
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java]: sorry, meant to write this on line 73

File: core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    workerInfoList.add(new BlockWorkerInfo(new WorkerNetAddress().setHost(localhostName)
+        .setRpcPort(PORT).setDataPort(PORT).setWebPort(PORT), Constants.GB, Constants.MB));
+    Assert.assertEquals(localhostName,
+        policy.getWorkerForNextBlock(workerInfoList, Constants.GB).getHost());
```
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java]: if the block needs 1GB , the local worker has (1GB - 1MB) available, and the remote worker has 1GB available, should the block be written to the remote worker?

Comment (gjhkael) [inline core/client/src/test/java/alluxio/client/file/policy/LocalFirstWithoutEvictionPolicyTest.java]: Because getAvailableBytes() is not return the actual available bytes of the worker, because the used bytes are updated only after a file commits. So, I need to reserve some space to get the available bytes. But, as you say, this policy would happened that the worker1 actually has 1GB, but the client do not get it to store 1GB data.

File: docs/_data/table/en/user-configuration.yml (hunk 3 above)
Comment (aaudiber) [inline docs/_data/table/en/user-configuration.yml]: Thanks for the link, this makes more sense now. Instead of using a ratio, do you think it makes more sense for the configuration to specify a certain number of MB to reserve instead of a ratio, and have it default to 0?

**Comment by gjhkael** [comment]
@aaudiber can you review it?

File: core/client/src/main/java/alluxio/client/block/BlockWorkerInfo.java [hunk 6]
```diff
@@ -56,6 +59,15 @@ public long getCapacityBytes() {
   }
 
   /**
+   * @return the available bytes of the worker
+   */
+  public long getAvailableBytes() {
+    mUserFileWriteCapacityReserved = Configuration
+        .getBytes(PropertyKey.USER_FILE_WRITE_CAPACITY_RESERVED_SIZE_BYTES);
+    return mCapacityBytes - mUsedBytes - mUserFileWriteCapacityReserved;
```
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/block/BlockWorkerInfo.java]: The space reservation logic is specific to `LocalFirstAvoidEvictionPolicy`, so we should put the logic there instead of modifying `BlockWorkerInfo`

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+import com.google.common.collect.Lists;
+
+import java.util.Collections;
+import java.util.List;
+
+import javax.annotation.concurrent.ThreadSafe;
+/**
+ * A policy that returns local host first, and if the local worker doesn't have enough availability,
+ * it randomly picks a worker from the active workers list for each block write.
+ * If No worker meets the demands, return local host.
```
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: Can you also specify the behavior of `USER_FILE_WRITE_CAPACITY_RESERVED_SIZE_BYTES`?

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 8]
```diff
@@ -252,6 +252,8 @@
       Name.USER_FILE_WORKER_CLIENT_POOL_GC_THRESHOLD_MS, 300 * Constants.SECOND_MS),
   USER_FILE_WRITE_LOCATION_POLICY(Name.USER_FILE_WRITE_LOCATION_POLICY,
       "alluxio.client.file.policy.LocalFirstPolicy"),
+  USER_FILE_WRITE_CAPACITY_RESERVED_SIZE_BYTES(
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/PropertyKey.java]: This property only modifies the new policy, so what do you think of naming it
`USER_FILE_WRITE_AVOID_EVICTION_POLICY_RESERVED_BYTES`?

File: docs/_data/table/en/user-configuration.yml [hunk 9]
```diff
@@ -30,6 +30,9 @@ alluxio.user.file.worker.client.threads:
   How many threads to use for file worker clients to read from workers.
 alluxio.user.file.write.location.policy.class:
   The default location policy for choosing workers for writing a file's blocks
+alluxio.user.file.write.capacity.reserved.size.bytes:
+  The portion of space reserced in worker when user use the LocalFirstAvoidEvictionPolicy class
```
Comment (aaudiber) [inline docs/_data/table/en/user-configuration.yml]: reserced -> reserved

Comment (gjhkael) [inline docs/_data/table/en/user-configuration.yml]: Sorry about this small mistake.

**Comment by gjhkael** [comment]
alluxio-bot, check this please

**Comment by gjhkael** [comment]
@aaudiber can you review it again, thanks.

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+ * USER_FILE_WRITE_AVOID_EVICTION_POLICY_RESERVED_BYTES is use to reserved some space of the worker
+ * to store the block, for the values mCapacityBytes minus mUsedBytes is not the available bytes.
+ */
+@ThreadSafe
+public class LocalFirstAvoidEvictionPolicy implements FileWriteLocationPolicy {
```
Comment (yupeng9) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: final

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+import java.util.Collections;
+import java.util.List;
+
+import javax.annotation.concurrent.ThreadSafe;
+/**
+ * A policy that returns local host first, and if the local worker doesn't have enough availability,
+ * it randomly picks a worker from the active workers list for each block write.
+ * If No worker meets the demands, return local host.
+ * USER_FILE_WRITE_AVOID_EVICTION_POLICY_RESERVED_BYTES is use to reserved some space of the worker
```
Comment (yupeng9) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: `is used to reserve`

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+    // otherwise randomly pick a worker that has enough availability
+    List<BlockWorkerInfo> shuffledWorkers = Lists.newArrayList(workerInfoList);
+    Collections.shuffle(shuffledWorkers);
+    for (BlockWorkerInfo workerInfo : shuffledWorkers) {
+      if (getAvailableBytes(workerInfo) >= blockSizeBytes) {
+        return workerInfo.getNetAddress();
+      }
+    }
+    return localWorkerNetAddress;
```
Comment (gpang) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: What if there was no worker on the local host? That means `localWorkerNetAddress` would have been null. Should `null` be returned, or should a random worker still be picked?

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  /**
+   * @param workerInfo BlockWorkerInfo of the worker
+   * @return the available bytes of the worker
```
Comment (gpang) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: Please comment how `USER_FILE_WRITE_AVOID_EVICTION_POLICY_RESERVED_BYTES` is used to compute the available bytes of the worker

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java (hunk 12 above)
Comment (gjhkael) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: If no worker on the local host, this policy will choice a random worker that have available capacity to store the block. If on worker have spare capacity, it should return null. This policy is similar to LocalFirstPolicy, but if local host do not have the spare capacity, it will select other worker instead of evict old block。

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java (hunk 13 above)
Comment (gjhkael) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: My original idea is use the variable mCapacityBytes minus mUsedBytes of class BlockWorkerInfo to get the valid Bytes of the worker. But yupeng9  tell me that this way does not actually get the available Bytes for the information of class BlockWorkerInfo is update only after a file is completely write. So, he suggest me to reserve some space to store the block.

Comment (gjhkael) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: I will comment this describe to this method. Thanks for you advise.

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java (hunk 11 above)
Comment (gjhkael) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: Fixed

File: core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java (hunk 12 above)
Comment (gjhkael) [inline core/client/src/main/java/alluxio/client/file/policy/LocalFirstAvoidEvictionPolicy.java]: @gpang After long deliberation, if no worker on the local host and have on worker have available capacity to store the block, it should pick a random worker. When no worker in this cluster, it should return null. Thanks for you remind, it make sense.

**Comment by gjhkael** [comment]
How to resolve conflicts?

**Comment by calvinjia** [comment]
@gjhkael You can run git merge master which will automatically merge the non-conflicting changes. You will then need to manually select which changes you want to keep (or if you want to keep parts of both) for the conflicting portions. Here is the github documentation: https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/

**Comment by gjhkael** [comment]
@calvinjia thanks for you help.
